`arrangement_polynomial`

danteluber opened the pull request:
Function which takes a matrix as an input, and returns defining polynomial for arrangement induced by columns.

I would not be surprised if this belongs somewhere other than PolyhedralFan.

danteluber committed:
hyperplane equation function

danteluber committed:
tesT

danteluber committed:
test

danteluber committed:
fix test

fingolfin commented on test/PolyhedralGeometry/polyhedral_fan.jl:
```diff
@@ -163,3 +163,9 @@ end
+@testset "Matrix to arrangement" begin
+  A = identity_matrix(QQ,3)
+  P,x = polynomial_ring(QQ,3)
+  @test matroid_to_arrangement(A) == x[1]*x[2]*x[3]
```
```suggestion
  @test matrix_to_arrangement(A) == x[1]*x[2]*x[3]
```

danteluber replied:
Thank you :D

danteluber committed:
fix test typo

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+## Hyperplane arrangements
+#################################################################################
+
+@doc raw"""
```
Docstrings should start with the function signature, indented like this (so it is formatted as code) :-)
```suggestion
@doc raw"""
    matrix_to_arrangement(A::MatElem{<: FieldElem})

```

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+
+@doc raw"""
+Let $A$ be a $d\times n$ matrix with entries from a field $\mathbb{F}$$.
+The collumns of $A$  are the normal vectors for a hyperplane arrangement
```
```suggestion
The columns of $A$  are the normal vectors for a hyperplane arrangement
```

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+
+Then we have $$\cup_{H_{i}\in\mathcal{A}}H_{i} = V(\Pi^{n}_{i=1}\alpha_{i}).$$
+
+Given some $A\in\mathbb{F}^{d\times n},$ return the product of the linear forms corresponding to the columns.
```
Shouldn't that rather be the start of the docstring? Then the rest is a more detailed explanation (or rather, perhaps, a justification why this return value here is of interest?!?

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
````diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+1 * (2*x1 + 3*x2 + 2*x3) * (7*x1 + 8*x2 + x3) * (x1 + 6*x2 + 10*x3) * (2*x1 + 4*x2 + 5*x3) * x3 * (3*x1 + x2 + 2*x3)
+```
+"""
+function matrix_to_arrangement(A::MatElem{<: FieldElem})
````
There are multiple issues with this function name (naming is one of the hardest problems, after all). I am not listing these to demotivate you, just to explain, I hope we can figure out a solution from there :-)
1. first off, we generally frown putting things into function names that are implied by the argument types. Here the argument is a matrix, so "matrix_to_" is kind of redundant
2. if one really needs the input type, I think the established pattern is more like `arrangement_from_matrix` (though instead of `matrix` a semantic description might be better... perhaps `arrangement_from_linear_forms`
3. one might just call this `arrangement`, dropping the prefix. But this function really returns a polynomial... Which I find really surprising, no matter which name variant listed so far is used. Of course if there is a common convention in your field that calls this polynomial an "arrangement", that would be fine. But if not, the next question might to ask "what do textbooks call this"?

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+    d = nrows(A)
+    n = ncols(A)
+    R = base_ring(A)
+    P,x = polynomial_ring(R,d)
```
```suggestion
    P,x = polynomial_ring(R,d; cached=false)
```
See also https://docs.oscar-system.org/dev/DeveloperDocumentation/caching/

You might then also consider adding a way for the user to specify a polynomial ring to use (e.g. `minpoly` and `charpoly` allow this by having an optional first argument of type `PolyRing`; others have a `parent` kwarg.

fingolfin commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,39 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+    R = base_ring(A)
+    P,x = polynomial_ring(R,d)
+    P = prod([sum([A[i,j]*x[i] for i in 1:d]) for j in 1:n])
+    return inv(unit(factor(P)))*P
```
Factoring the polynomial just to get this unit out seems expensive to me. Is it really necessary, or just for beautification?

fingolfin reviewed (commented):
Thank you for your contribution, much appreciated! A few notes based on a quick look on your PR.

danteluber committed:
implementing suggested changes

danteluber committed:
typo

danteluber committed:
forgot to change docs

lgoettgens commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
````diff
@@ -303,3 +303,45 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+Then we have $$\cup_{H_{i}\in\mathcal{A}}H_{i} = V(\Pi^{n}_{i=1}\alpha_{i}).$$
+
+# Example
+```julia> A = matrix(QQ,[1 0 2 1//2 3 7;2 0 3 3 1 8;5//2 1 2 5 2 1])
````
```suggestion
```jldoctest
julia> A = matrix(QQ,[1 0 2 1//2 3 7;2 0 3 3 1 8;5//2 1 2 5 2 1])
```
some formatting fix

danteluber committed:
docs

danteluber committed:
typo

danteluber committed:
errors in docs

danteluber committed:
test

danteluber committed:
formatting in test

fingolfin commented:
CI tests fail

danteluber committed:
typo in test

danteluber commented:
> CI tests fail

Error in tests addressed. Seems to be fine now.

fingolfin commented:
It now complains about formatting issues *sigh*. According to the CI logs:

> The files
["src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl"]
were not properly formatted. You can fix the formatting by running
```
using JuliaFormatter
format("src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl")
```


Perhaps you can give it a try? Otherwise I'd be tempted to just ignore or even disable that CI check, as per #3805

lkastner commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -303,3 +303,52 @@ function Base.:*(PF1::PolyhedralFan{QQFieldElem}, PF2::PolyhedralFan{QQFieldElem
+    @req base_ring(A) == coefficient_ring(ring) "entries of input matrix must be coefficients from ring"
+    x = gens(ring)
+    n = ncols(A)
+    return prod([sum([A[i,j]*x[i] for i in 1:length(x)]) for j in 1:n])
```
```suggestion
    return prod(transpose(A)*x)
```
Is that the same?

lkastner committed:
Update src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl

lkastner committed:
PolyhedralGeometry: Enhance arrangement_polynomial

- Allow some more signatures
- Use rows instead of cols
- Optional parameter for using cols instead of rows

lkastner committed:
PolyhedralGeometry: Some more signatures for arrangement_polynomial

benlorenz commented on test/PolyhedralGeometry/polyhedral_fan.jl:
```diff
@@ -163,3 +163,28 @@ end
+  RAvv = parent(apAvv)
+  apAM = arrangement_polynomial(RAvv, AM)
+  @test nvars(RAvv) == 3
+  @test apAvv = apAM
```
```suggestion
  @test apAvv == apAM
```

benlorenz committed:
Update test/PolyhedralGeometry/polyhedral_fan.jl
